robot_calibration_3d/src/joint_state_publisher.cpp

```diff
@@ -0,0 +1,95 @@
+#include "robot_calibration_3d/joint_state_publisher.h"
+
+JointStatePublisher::JointStatePublisher(ros::NodeHandle nh) : nh_(nh)
+{
+  // TODO: retrieve this value from the URDF rather than hardcode
+  num_joints_ = 6;
+
+  // initialising publisher
+  joint_states_pub_ = nh_.advertise<sensor_msgs::JointState>("joint_states", 20);
+
+  // retrieving state update parameters from param server
+  nh_.param("robot_joint_angles/num_state_changes", num_state_changes_, 50);
+  nh_.param("robot_joint_angles/state_update_interval", state_update_interval_float_, 3.0);
+
+  // ensure the update interval is positive
+  if (state_update_interval_float_ <= 0)
+  {
+    ROS_WARN("Non-positive state update interval specified (%f), interval updated to default (%f).",
+             state_update_interval_float_, 3.0);
+    state_update_interval_float_ = 3.0;
+  }
+
+  // getting update interval as a ros::Duration rather than a double eases calculation
+  ros::Duration state_update_interval_duration(state_update_interval_float_);
+
+  // set rate at which messages are sent, regardless of when state is updated
+  ros::Rate loop_rate_(10);
+
+  // seed number generator to ensure randomness
+  srand(time(0));
+
+  // populate the joint name array of the message, assuming the URDF uses the naming convention that the ith joint is
+  // named "joint_i"
+  // TODO: retrieve joint names directly from URDF, rather than assume the naming convention used below
+  for (int i = 1; i <= num_joints_; i++)
+    joint_states_message_.name.push_back("joint_" + std::to_string(i));
+
+  // initialise the joint state array of the message with random values for first update
+  // TODO: retrieve angle limits from URDF, rather than hardcoding limits as is done in this function
+  updateJointStateMessage(&joint_states_message_);
+
+  // set time of update after this first update
+  ros::Time last_update_time = ros::Time::now();
+
+  // start counting updates
+  int count = 1;
+
+  while (ros::ok())
+  {
+    // if enough time has passed since last update and number of desired state changes has not yet been met, update
```

Comment:
Encapsulate this logic into a public function, publishJointStateMessage, which should be called on every loop in your node script. Have this function check if it is time to update the joints, update the joints if it's time to do that and publish the joint states.